### ISSUE TITLE
Filter missing if wrong type passed in

### ISSUE THREAD

**Issue by ztone** [aspnet/EntityFramework#1858]
When I accidentally passed in an int as a filter to a column that is a smallint/short then the SQL generated became skewed without any kind of warning.

Also tried <code>Disdatefilter  >=</code> and just using <code>Equals()</code> like in this example.

Here is the model/context
```c#
public class Context : DbContext
{
    public DbSet<Data> Data { get; set; }

    protected override void OnConfiguring(DbContextOptions options)
    {
        options.UseSqlServer(@"..");
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        base.OnModelCreating(modelBuilder);
        modelBuilder.Entity<Data>().Key(sd => new { sd.IdType, sd.Id });
    }
}

public class Data
{
        public short Provider { get; set; }
        public short DisdateFilter { get; set; }
        public long IdType{ get; set; }
        public byte Id { get; set; }
        public bool Excluded{ get; set; }
}
```
The linq query used
```c#
int start = 1403;
using (var ctx = new Context ())
{
   var query = ctx.Data.Where(x => x.DisdateFilter.Equals(start) && x.Provider.Equals(612))
      .Select(x => new { x.IdType });
}
```

Here is the skewed SQL that it produced.  Two things I notice here.  One is of course that the DisdateFilter is missing from the where clause.  But also that the SQL statement is selecting the filtered columns.
```sql
exec sp_executesql N'SELECT [x].[DisdateFilter], [x].[Provider], [x].[IdType]
FROM [Data] AS [x]
WHERE [x].[Provider] = @p0',N'@p0 smallint',@p0=612
```
When using the correct type the SQL is correct.
```sql
exec sp_executesql N'SELECT [x].[IdType]
FROM [Data] AS [x]
WHERE ([x].[DisdateFilter] = @p0 AND [x].[Provider] = @p1)',N'@p0 smallint,@p1 smallint',@p0=1403,@p1=612
```

**Comment by maumar** [comment]
Investigation results: The problem in the way parameter is evaluated. For Equals method with not-matching types Equals(object, object) is being used. Parameter is wrapped in a convert and when we evaluate it, it comes out as object. Later on when we encounter parameter of type object we don't know how to translate it to SQL (in general), so this part of the predicate is done on the client.

**Comment by maumar** [comment]
Filed https://github.com/aspnet/EntityFramework/issues/1927 - related to this, it should be fixed in the next beta.

**Comment by maumar** [comment]
We will translate Equals(object) to FALSE on the server, if the types of operands are different as well as log a warning that this is probably not what the user intended. This is consistent with CLR behavior which will always return false in this case (per http://blog.coverity.com/2014/01/13/inconsistent-equality).

If the other overload of Equals is used, or the types are the same, we will translate to == operator.

**Comment by maumar** [comment]
Fixed in 517fc188c64718fc51f9b3e38acf12ac5e43a722
